include/alpaka/mem/view/Accessor.hpp

```diff
@@ -0,0 +1,433 @@
+        struct BuildAccessModeList;
+
+        template<typename TAccessMode>
+        struct BuildAccessModeList<TAccessMode>
+        {
+            using type = TAccessMode;
+        };
+
+        template<typename TAccessMode1, typename TAccessMode2, typename... TAccessModes>
+        struct BuildAccessModeList<TAccessMode1, TAccessMode2, TAccessModes...>
+        {
+            using type = std::tuple<TAccessMode1, TAccessMode2, TAccessModes...>;
+        };
+
+        template<typename... TAccessModes, typename TBuf, std::size_t... TPitchIs, std::size_t... TExtentIs>
+        ALPAKA_FN_HOST_ACC auto buildAccessor(
+            TBuf&& buffer,
+            std::index_sequence<TPitchIs...>,
+            std::index_sequence<TExtentIs...>)
+        {
+            using DBuf = std::decay_t<TBuf>;
+            using TBufferIdx = Idx<DBuf>;
+            constexpr auto dim = Dim<DBuf>::value;
+            auto p = getPtrNative(buffer);
+            using AccessModeList = typename BuildAccessModeList<TAccessModes...>::type;
+            return Accessor<decltype(p), Elem<DBuf>, TBufferIdx, dim, AccessModeList>{
+                p,
+                getPitchBytes<TPitchIs + 1>(buffer)...,
+                {extent::getExtent<TExtentIs>(buffer)...}};
+        }
+
+        template<typename T>
+#ifdef __cpp_inline_variables
+        inline
+#endif
+            constexpr bool isAccessor
+            = false;
+
+        template<typename TMemoryHandle, typename TElem, typename TBufferIdx, std::size_t Dim, typename TAccessModes>
+#ifdef __cpp_inline_variables
+        inline
+#endif
+            constexpr bool isAccessor<Accessor<TMemoryHandle, TElem, TBufferIdx, Dim, TAccessModes>> = true;
+    } // namespace internal
+
+    //! Creates an accessor for the given memory object (view or buffer) using the specified access modes.
+    template<
+        typename... TAccessModes,
+        typename TBuf,
+        typename = std::enable_if_t<!internal::isAccessor<std::decay_t<TBuf>>>>
+    ALPAKA_FN_HOST_ACC auto accessWith(TBuf&& buffer)
+    {
+        using Dim = Dim<std::decay_t<TBuf>>;
+        return internal::buildAccessor<TAccessModes...>(
+            std::forward<TBuf>(buffer),
+            std::make_index_sequence<Dim::value - 1>{},
+            std::make_index_sequence<Dim::value>{});
+    }
+
+    //! Constrains an existing accessor with multiple access modes to the specified access modes.
+    // TODO: currently only allows constraining down to 1 access mode
+    template<
+        typename TNewAccessMode,
+        typename TMemoryHandle,
+        typename TElem,
+        typename TBufferIdx,
+        std::size_t TDim,
+        typename... TPrevAccessModes>
+    ALPAKA_FN_HOST_ACC auto accessWith(
+        const Accessor<TMemoryHandle, TElem, TBufferIdx, TDim, std::tuple<TPrevAccessModes...>>& acc)
+    {
+        static_assert(
+            boost::mp11::mp_contains<std::tuple<TPrevAccessModes...>, TNewAccessMode>::value,
```

Comment:
I could not find the equivalent meta function in alpaka so I used mp11. Please tell me if it already exists!